content/dotnet-template-azure-iot-edge-module/CSharp/ModuleBackgroundService.cs

```diff
@@ -0,0 +1,50 @@
+﻿using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Client.Transport.Mqtt;
+using System.Text;
+
+namespace SampleModule;
+
+internal class ModuleBackgroundService : BackgroundService
+{
+    private int _counter;
+    private ModuleClient? _moduleClient;
+    private readonly ILogger<ModuleBackgroundService> _logger;
+
+    public ModuleBackgroundService(ILogger<ModuleBackgroundService> logger) => _logger = logger;
+
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        MqttTransportSettings mqttSetting = new(TransportType.Mqtt_Tcp_Only);
+        ITransportSettings[] settings = { mqttSetting };
+
+        // Open a connection to the Edge runtime
+        _moduleClient = await ModuleClient.CreateFromEnvironmentAsync(settings);
+        await _moduleClient.OpenAsync(stoppingToken);
+        _logger.LogInformation("IoT Hub module client initialized.");
+
+        // Register callback to be called when a message is received by the module
+        await _moduleClient.SetInputMessageHandlerAsync("input1", PipeMessage, null, stoppingToken);
+    }
+
+    async Task<MessageResponse> PipeMessage(Message message, object userContext)
+    {
+        int counterValue = Interlocked.Increment(ref _counter);
+
+        byte[] messageBytes = message.GetBytes();
+        string messageString = Encoding.UTF8.GetString(messageBytes);
+        _logger.LogInformation("Received message: {counterValue}, Body: [{messageString}]", counterValue, messageString);
+
+        if (!string.IsNullOrEmpty(messageString))
+        {
+            using var pipeMessage = new Message(messageBytes);
+            foreach (var prop in message.Properties)
+            {
+                pipeMessage.Properties.Add(prop.Key, prop.Value);
+            }
+            await _moduleClient!.SendEventAsync("output1", pipeMessage);
```

Comment:
What is the `!` symbol here? I'm unfamiliar with it.

Reply:
[null forgiving](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) to avoid a code analysis warning

Reply:
Cool. Because we have the `<Nullable>enable</Nullable>` on in the project. I haven't played with that too much.